Realm/Realm/Native/SyncSocketProvider.cs

```diff
@@ -0,0 +1,207 @@
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate void provider_dispose(IntPtr managed_provider);
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate IntPtr create_timer(IntPtr socket_provider, UInt64 delay_miliseconds, IntPtr native_callback);
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate void cancel_timer(IntPtr timer);
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate IntPtr websocket_connect(IntPtr socket_provider, IntPtr observer, Endpoint endpoint);
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate void websocket_write(IntPtr managed_websocket, BinaryValue data, IntPtr native_callback);
+
+            [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+            public delegate void websocket_close(IntPtr managed_websocket);
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_websocket_install_callbacks", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void install_callbacks(post_work post, provider_dispose dispose, create_timer create_timer, cancel_timer cancel_timer, websocket_connect connect, websocket_write write, websocket_close close);
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_websocket_run_callback", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void run_callback(IntPtr native_callback, ErrorCode result, StringValue reason);
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_websocket_observer_connected_handler", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void observer_connected_handler(IntPtr observer, StringValue protocol);
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_websocket_observer_error_handler", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void observer_error_handler(IntPtr observer);
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_websocket_observer_binary_message_received", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void observer_binary_message_received(IntPtr observer, BinaryValue data);
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_websocket_observer_closed_handler", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void observer_closed_handler(IntPtr observer, WebSocketCloseStatus status, StringValue reason);
+        }
+
+        private static void PostWork(IntPtr managed_provider, IntPtr nativeCallback)
+        {
+            var provider = (SyncSocketProvider)GCHandle.FromIntPtr(managed_provider).Target;
+            provider._workQueue.Writer.TryWrite(new EventLoopWork(nativeCallback, Status.OK));
```

Comment:
It will return false if the channel is closed, and it's closed when the Sync Client tears down the `SyncSocketProvider` instance and at that time it's impossible to call `PostWork()`. That is to say, `TryWrite()` should never return false here.

Reply:
Should we assert on this?

Reply:
I changed it to `await WriteAsync()` instead of calling `TryWrite()`.